test/e2e/config_test.go

```diff
@@ -118,3 +124,960 @@ func assertOperatorCondition(t *testing.T, conditionType configv1.ClusterStatusC
+		},
+		{
+			name: "assert remote write url value in set in CR",
+			f:    assertRemoteWriteWasSet(f.UserWorkloadMonitoringNs, crName, "https://test.remotewrite.com/api/write"),
+		},
+	} {
+		if ok := t.Run(scenario.name, scenario.f); !ok {
+			t.Fatalf("scenario %q failed", scenario.name)
+		}
+	}
+}
+
+func TestUserWorkloadMonitorThanosRulerConfig(t *testing.T) {
+	const (
+		component       = "thanos-ruler"
+		pvcClaimName    = "thanos-ruler-user-workload-data-thanos-ruler-user-workload-0"
+		statefulsetName = "thanos-ruler-user-workload"
+		cpu             = "1m"
+		mem             = "3Mi"
+		storage         = "2Gi"
+	)
+
+	cm := &v1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "cluster-monitoring-config",
+			Namespace: f.Ns,
+		},
+		Data: map[string]string{
+			"config.yaml": `enableUserWorkload: true
+`,
+		},
+	}
+
+	uwmCM := &v1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "user-workload-monitoring-config",
+			Namespace: f.UserWorkloadMonitoringNs,
+		},
+		Data: map[string]string{
+			"config.yaml": fmt.Sprintf(`thanosRuler:
+  logLevel: debug
+  tolerations:
+    - operator: "Exists"
+  volumeClaimTemplate:
+    spec:
+      resources:
+        requests:
+          storage: %s
+  resources:
+    requests:
+      cpu: %s
+      memory: %s
+`, storage, cpu, mem),
+		},
+	}
+
+	if err := f.OperatorClient.CreateOrUpdateConfigMap(cm); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := f.OperatorClient.CreateOrUpdateConfigMap(uwmCM); err != nil {
+		t.Fatal(err)
+	}
+	for _, scenario := range []struct {
+		name string
+		f    func(*testing.T)
+	}{
+		{
+			name: "assert that PVC is created and ss rolled out",
+			f: assertVolumeClaimsConfigAndRollout(rolloutParams{
+				component:       component,
+				namespace:       f.UserWorkloadMonitoringNs,
+				claimName:       pvcClaimName,
+				statefulSetName: statefulsetName,
+			}),
+		},
+		{
+			name: "assert that pod config is correct",
+			f: assertPodConfiguration(
+				podConfigParams{
+					component:     component,
+					namespace:     f.UserWorkloadMonitoringNs,
+					labelSelector: "app.kubernetes.io/name=thanos-ruler",
+				},
+				[]podAssertionCB{
+					expectCatchAllToleration(),
+					expectMatchingRequests("*", component, mem, cpu),
+				},
+			),
+		},
+		{
+			name: "assert assets are deleted when user workload monitoring is disabled",
+			f:    assertDeletedUserWorkloadAssets(cm),
+		},
+	} {
+		if ok := t.Run(scenario.name, scenario.f); !ok {
+			t.Fatalf("scenario %q failed", scenario.name)
+		}
+	}
+}
+
+type deploymentRolloutParams struct {
+	namespace, name string
+}
+
+func assertDeploymentRollout(params deploymentRolloutParams) func(*testing.T) {
+	return func(t *testing.T) {
+		err := f.OperatorClient.WaitForDeploymentRollout(&appsv1.Deployment{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      params.name,
+				Namespace: params.namespace,
+			},
+		})
+		if err != nil {
+			t.Fatal(err)
```

Comment:
If we call `Fatal` here, `Fatalf` from the caller site will never be invoked?

```
		if ok := t.Run(scenario.name, scenario.f); !ok {
			t.Fatalf("scenario %q failed", scenario.name)
		}
```

Maybe it would be clearer if this function doesn't take `*testing.T` as input and  returns only an error? This way we don't have to think about who should fail? 

```
func assertDeploymentRollout(params deploymentRolloutParams) func() error {
```

Reply:
Just coming back on this one with some context. A lot of the tests in this package already use table driven tests with the format 
```go
	for _, scenario := range []struct {
		name string
		f    func(*testing.T)
	}
```

so while I agree with your point, I structured it like this for consistency across tests in the e2e test suite. So I think we have three options here

1. Leave this as is for consistency across the package
2. Refactor it as suggested and update the other tests in the package also (most work)
3. Remove it from the table driven test format and check this one (refactored as per your suggestion) explicitly before running the table driven tests in existing format

wdyt?

Reply:
I get your point. What about not checking `t.Run()`'s returned value since it's the responsibility of the scenario function to `Fail`?